libs/SerialNumberKit/source/SerialNumberKit.cpp

```diff
@@ -7,34 +7,66 @@
 #include <cstring>
 
 #include "CastUtils.h"
+#include "FileManagerKit.h"
 
 using namespace leka;
 
 auto SerialNumberKit::getSerialNumber() -> std::array<uint8_t, SN_SIZE> &
 {
-	auto prefix = utils::cast::from_c_string_to_uint8_t_array("LK-22xx-");
-	std::copy(prefix.begin(), prefix.end(), serial_number.begin());
+	auto prefix_size	   = setPrefix(serial_number);
+	auto date_of_test_size = setDateOfTest(serial_number, prefix_size);
+	auto number_of_digits  = static_cast<uint8_t>(SN_SIZE - (prefix_size + date_of_test_size + sizeof('\0')));
 
-	std::array<char, SN_SIZE - std::size(prefix) + 1> mcu_id_array {};
-	auto mcu_id = _mcu.getID();
-	snprintf(mcu_id_array.begin(), std::size(mcu_id_array), "%08lX%08lX%08lX", mcu_id.front, mcu_id.middle,
-			 mcu_id.back);
-
-	std::memcpy(serial_number.begin() + std::size(prefix) - 1, &mcu_id_array, std::size(mcu_id_array));
+	setMCUID(serial_number, prefix_size + date_of_test_size, number_of_digits);
 
 	return serial_number;
 }
 
 auto SerialNumberKit::getShortSerialNumber() -> std::array<uint8_t, SHORT_SN_SIZE> &
 {
-	auto prefix = utils::cast::from_c_string_to_uint8_t_array("LK-22xx-");
-	std::copy(prefix.begin(), prefix.end(), short_serial_number.begin());
-
-	std::array<char, SHORT_SN_SIZE - std::size(prefix) + 1> mcu_id_array {};
-	auto mcu_id = _mcu.getID();
-	snprintf(mcu_id_array.begin(), std::size(mcu_id_array), "%08lX", mcu_id.front);
+	auto prefix_size	   = setPrefix(short_serial_number);
+	auto date_of_test_size = setDateOfTest(short_serial_number, prefix_size);
+	auto number_of_digits  = static_cast<uint8_t>(SHORT_SN_SIZE - (prefix_size + date_of_test_size + sizeof('\0')));
 
-	std::memcpy(short_serial_number.begin() + std::size(prefix) - 1, &mcu_id_array, std::size(mcu_id_array));
+	setMCUID(short_serial_number, prefix_size + date_of_test_size, number_of_digits);
 
 	return short_serial_number;
 }
+
+auto SerialNumberKit::setPrefix(std::span<uint8_t> partial_serial_number) const -> uint32_t
+{
+	auto prefix = utils::cast::from_c_string_to_uint8_t_array("LK-");
+	std::copy(prefix.begin(), prefix.end(), partial_serial_number.begin());
+
+	return std::size(prefix) - 1;
+}
+
+auto SerialNumberKit::setDateOfTest(std::span<uint8_t> partial_serial_number, uint32_t offset) const -> uint32_t
+{
+	auto date_of_test = std::array<uint8_t, 4> {};
+	date_of_test.fill('X');
+
+	if (auto file = FileManagerKit::File {date_of_test_path}; file.is_open()) {
+		file.read(date_of_test);
+		file.close();
+	}
+
+	std::copy(date_of_test.begin(), date_of_test.end(), partial_serial_number.begin() + offset);
+
+	return std::size(date_of_test);
```

Comment:
A bit tricky but yes. 

In fact, unit test check that the correct size is returned (without considering if it's `-1` or not)

The difference in implementation is that in `setPrefix` the array is generated from a c_string with null terminated since `setDateOfTest` read the 4 first bytes of the file using an array

Reply:
Prefix with array {'L', 'K', '-'}